File: 0172. Factorial Trailing Zeroes/172.factorial-trailing-zeroes.java
//Given an integer n, return the number of trailing zeroes in n!.
//
// Note that n! = n * (n - 1) * (n - 2) * ... * 3 * 2 * 1.
//
//
// Example 1:
//
//
//Input: n = 3
//Output: 0
//Explanation: 3! = 6, no trailing zero.
//
//
// Example 2:
//
//
//Input: n = 5
//Output: 1
//Explanation: 5! = 120, one trailing zero.
//
//
// Example 3:
//
//
//Input: n = 0
//Output: 0
//
//
//
// Constraints:
//
//
// 0 <= n <= 10⁴
//
//
//
// Follow up: Could you write a solution that works in logarithmic time
//complexity?
//
// Related Topics Math 👍 3337 👎 1973


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int trailingZeroes(int n) {
        int res = 0;
        int divider = 5;
        while (n >= divider) {
            res += n / divider;
            divider *= 5;
        }
        return res;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 0172. Factorial Trailing Zeroes/172.factorial-trailing-zeroes.go
package leetcode

/*
 * @lc app=leetcode id=172 lang=golang
 *
 * [172] Factorial Trailing Zeroes
 */

// @lc code=start
func trailingZeroes(n int) int {
	// n < 10^4
	// 5   的倍數: 5, 10, 15, 20, 25, 30, 35    -> 提供1個 5
	// 25  的倍數: 25, 50, 75, ...              -> 提供2個 5
	// 125 的倍數: 125, 250, 375, ...           -> 提供3個 5

	// E.g. n = 10
	// 10 / 5 = 2，會有 2 個 5
	// 總共 2 個
	// E.g. n = 30
	// 30 / 5  = 6，會有 6 個 5
	// 30 / 25 = 1，會有 1 個 5 (25 會提供 2 個，所以要補回來)
	// 總共 7 個
	// E.g. n = 130
	// 130 / 5 = 26，會有 26 個 5
	// 130 / 25 = 5，會有 5 個 5 (25 會提供 2 個，所以要補回來)
	// 130 / 125 = 1，會有 1 個 5 (125 會提供 3 個，所以要補回來)
	// 總共 32 個
	res := 0
	divisor := 5
	for n >= divisor {
		res += n / divisor
		divisor *= 5
	}
	return res
}

// @lc code=end
